### DISCUSSION TITLE
IR Remote library that works with 0 and 1 series?

### DISCUSSION THREAD

**Discussion by intakenick** [SpenceKonde/megaTinyCore#784] [Library discussion]
I've developed a product around the 1614 and figured hooking up a receiver for a NEC remote would be trivial but there appears to be no compatible libraries. Can anyone point me towards one?

**Comment by grandaspanna** [comment]
Not a complete answer, but I think it's pretty close. This library: https://github.com/Arduino-IRremote/Arduino-IRremote knows about megatiny's, but not the 1614 specifically. So, most of the work done supporting the timers (timerA at least) is already done. I was able to compile the simple receiver example for a 1616, which is darn close to a 1614. 

Playing around with some #defines should get you there. There's a few bits in here that control which timer to use, depending on the architecture: https://github.com/Arduino-IRremote/Arduino-IRremote/blob/master/src/private/IRTimer.hpp

Seems to specifically call out a few chips:
```
#elif defined(__AVR_ATtiny1616__)  || defined(__AVR_ATtiny3216__) || defined(__AVR_ATtiny3217__) // e.g. TinyCore boards
#  if !defined(IR_USE_AVR_TIMER_A) && !defined(IR_USE_AVR_TIMER_D)
#define IR_USE_AVR_TIMER_A // use this if you use MegaTinyCore, Tone is on TCB and millis() on TCD
//#define IR_USE_AVR_TIMER_D // use this if you use TinyCore
#  endif
```
I suspect there's a better identifier to use that is a bit more generic. Would be good idea to raise it as an issue, because otherwise, the library looks handy.

**Reply by grandaspanna** [reply]
If just add this at the top of the example:
`#define IR_USE_AVR_TIMER_A
`
The code compiles for a 1614. Haven't tried running it...

**Reply by intakenick** [reply]
Awesome, I'll give this a shot! Also probably a good time for me to brush up on timers. Thank you

**Comment by SpenceKonde** [comment]
You can test for megatinycore with #if defined(MEGATINYCORE), test the series of parts with #if (MEGATINYCORE_SERIES== ... ) which should be needed if you're fixing the library (Thank you kindly if that's what you're doing, get the bloody fix PR'ed in)

Test for MILLIS_USE_TIMERA0, MILLIS_USE_TIMERD0, etc to see which timer they're using for millis. Defaults to TCA0 on 0-series so that there's one free type B timer, allowing tone() or any other timer requiring thing, TCD0, on the 1-series (on the logic that that timer a byzantine horror show - albeit an extremely powerful one - that nobody who cared about their own sanity would touch it with a 39.5 foot pole - so mad props to the library author! And 2-series defaults TCB1 because got ripped off and don't have the type D timer (I know why  - it's because they only appeared in the 1-series to test out for Dx-series.). 

(And basic props to anyone who knows where the 39.5 foot pole line came from - it's a lyric from a song associated with a classic book and movie, but few have committed that jeremiad to memory. Its got some good, if nonspecific, zingers, and a lot of infantile ones, but best of all it works with my first name in place of the villian's name (keeping the Mr. which precedes it). Though I think most of the insults better fit timer D than they do me.)

**start of offtopic and at times very controversial rambling**
Speaking of insults and song lyrics, I just stumbles across another gem, from diamond head back in '80 - "Am I evil? Yes I am!" - inspired stuff you I can really relate to - sadly metallica covered it (poorly might I add, if I wanted to use gratuitous profanity I'd say they covered it in shit) and the song is associated with them, and poor diamond head was a 0-hit wonder instead of a 1 hit wonder. There are a lot of 1-hit and 2-hit wonders from the UK in the decade surrounding my birth, Eazy Money - Telephone Man is nothing less than a masterpiece (and for once the record company's post-recording edits actually improved the song), Bitch's Sin - Strangers on the shore (only the demo version is good, but man it doesn't get much better than that, they lost the good guitarist and plunged down the drain almost immediately. if it doesn't open with a tapping solo you got the bad one. good one was on the heavy metal heroes compilation), and Twisted Ace with I won't Surrender (where he vow to not become one of the forgotten heavy metal brigade, which he promptly did as evidenced by his being the only band I had to look up the name of here), and This Fire Inside, which is a lovely mondegreen which suggests alternate lyrics inspired by that "I know an old lady who swallowed a fly"... ("This spider inside, this spider inside, swallowed that thing just to eat the fly, the spider inside oh the spider inside, I guess I'll eat livestock and die") 

That was all of course before, by demonstrating that competence at the guitar wasn't required for success, Nirvana ruined music (certainly everything from the USA), which only began to recover two decades after his murder; I'm thankful for Ms. Love prematurely ending his career and impressed she got away with it since it's so bloody obvious how that went down - though of course such a severely botched investigation that lets a murderer off is always unsetting, if she didn't do him in, he could have lasted until the fent scourge dropped him with an OD (and with that kind of habit you know it would have) so we could have still been 15 years away from musicians experimenting with learning to play guitar before they got on stage. And by then I'd be too old for metal shows. I didn't mind nirvana's music, he had lyrical and compositional finesse that his contemporaries didn't copy, but i detest what he did to music. Before him, a lead guitarist in a rock or metal band could be counted on to know how to shred at least passably. Not so after him, and really what's the point of listening to a song if there's not at least one killer solo in there? Preferably a minisolo elsewhere in the song too or some epic twin guitar action, or solos where a kb and guitar trade off like CoB did so brilliantly (RIP - Resurrect in Power. Alcohol is poison; I didn't drink before he passed away either but now not only do I not care forit, I hold a personal grudge). Good example is their cover of Don't Stop at the Top, Those two were so good it takes a bunch of listens to pinpoint where exactly they switch off in the solo, though you hear it the first time through that they were doing so). But we're getting of topic aren't we....? Where were we?


**back on topic**
The 1614 is one of the golden 1-series or 1.5-series, or whatever you want to call them, so it does have both a TCB0 and TCB1 (as well as a second ADC, and some other really sick extras).  Which gives you a ton of flexibility (remember library should be prepared for any timer to be selected for millis as that's just a submenu, and should be prepared to respond to any valid millis timer being selected - if it can't do what it needs with a TCB (really? I would think the TCB's would be 100% the best timers for that). I'd use this timer selection logic (and I realize now you don't need to test for the series, just which peripherals are there and that either DXCORE or MEGATINYCORE is defined)
```
#if defined(TCD0) && !defined(MILLIS_USE_TIMERD0) // also ideally want some way to tell it not to use TCD0 if you want to use that timer, but you can't propagate the the #define from the sketch to the c or cpp file where you will need it, only to the .h file that the SKETCH sees, while the one the library c files will not see the define, and the result will probably be a shitshow. 
    //use TCD0
    takeOverTCD0();
#elif defined(TCB1) && !defined MILLIS_USE_TIMERB1) 
    //use TCB1
#elif defined(TCB0) && !defined MILLIS_USE_TIMERB0)
    //use TCB0
// To support DX and future EA, you want to repeat same test for all the TCBs 4, 3, 2, 1, 0 - in descending order, though TCB2 will usually be doing millis. and an analogous test here for TCA1 which the bigger chips and all EA's will have. Actually maybe I'd make it the top choice if TCA1 is defined but TCD0 isn't (means its an EA-series). Certainly if _AVR_PINCOUNT < 48 and there's a TCA1, it's probably the least useful timer, in which case there's no way to get 6 channel output from TCA1)
#elif !defined(MILLIS_USE_TIMERA0) 
    //use TCA0
    takeOverTCA0();
#else  // shit here to stop trying to turn preprocessor directives into issue links!
   #errror "Cannot find a usable timer" 
#endif //fuck you again github
```
That should should figure out a non-stupid timer to use. But I can also see having to do a number of copies with different names that will prefer different timers because of the inability to propagate preprocessor directives to the library, which is in a different compilation unit. And IMHO, don't bother supporting that other tinyCore, he lost the war with megaTinyCore. I moved fast once he had effectively declared himself to be a threat, and by the time he fortified one part of his core against my assaults I'd rolled a battalion of tanks into two others. I mean, I originally offered to to collaborate, and support the (strictly worse) pinmappings his hardware was selling as first class citizens of megaTinyCore, but he rebuked me *shrug*. I don't like competition with my cores; I take it very personally. And that's why I strive to make sure mine are always the best. And if war was what he wanted, war was what he'd get. And he lost. I burned his towns, looted his cities and sowed his fields with salt - he fled, and stopped developing. In these times only those who have lost their way venture into the ruins of tinycore. So you've just gotta cover DxCore and megaTinyCore (which are extremely similar - they share an incredible amount of code), and MegaCoreX (where you can count on millis always using TCB2 because he doesn't give users a menu to select the timer for millis- i think it's safe to just use TCB1 all the time on that one.

**Reply by intakenick** [reply]
Well this is about the best answer I could have asked for. I appreciate that anecdote it certainly went in directions I was not expecting (and for the most part I agree). I'll give this solution a shot

**Comment by 6v6gt-duino** [comment]
Maybe this is too late for you, but if you are interested only in reading the NEC protocol, this may help. It does not use timers explicitly, instead it uses an external interrupt. https://forum.arduino.cc/t/lightweight-arduino-ir-library-for-nec-remote-control-devices/595247 
I developed it for an ATTINY85 at the time when the what was once the  Ken Shirriff Arduino IR Remote library did not support it.

**Comment by SpenceKonde** [comment]
I would think that an IR-Remote library could be made insanely super efficient on a modern AVR. It's it like.... an XX kHz carrier, modulated by data at some much slower data rate? As in the kind of task that makes the CCL logic blocks climb out of the chip and line up begging you to use them?
